[benchmarks] Add repeat benchmark

kevinpschaaf opened the pull request:
Adds a benchmark for the `repeat` directive (dug out of an old stash, used during the original repeat implementation).

By default, it runs a series of different operations on the list. The following `performance.measure` entries are collected for each operation:

* `render` - initial render (1000 items)
* `nop` - no-op render with same data
* `replace` - replaces the entire list with new data
* `add` - adds a set of contiguous items in the middle of the list
* `addMirror` - adds items in two spots in the list equidistant from each end
* `remove` - removes items from the middle of the list
* `removeMirror` - removes items from two spots in the list equidistant from each end
* `reverse`- reverses the list
* `swapOne` - swaps one item in the list
* `swapMany` - swaps meny items in the list
* `moveForwardOne` - moves one item forward in the list
* `moveForwardMany` - moves a set of contiguous items forward in the list
* `moveBackwardOne` - moves one item backwards in the list
* `moveBackwardMany` - moves a set of contiguous items backward in the list
* `shuffle` - does a random shuffle of the list
* `update` - total of everything except `render`
* `total` - total time for all operations

Each operation is done in a loop a hand-tuned number of times, such that each operation takes _roughly_ the same amount of time, to avoid one operation swamping the others.

The default `lit-html-repeat` benchmark on CI only reports `render` and `update`. 

There is a `lit-html-repeat-full` tachometer configuration that reports a measure for each individual operation. That would end up making the PR comment huge, so I was trying to find a good way that could be enabled/triggered on-demand, i.e. just for PR's that might affect `repeat` specifically, but didn't come to a great solution. I don't think we want to trigger running the benchmarks on every label or PR body change, but whatever event we use needs to happen after the change in "state" we use to note that the PR should have the benchmarks run. TBD.

sorvell commented:
Other tests that might be good but this might be overkill
* `add random`: adds a set of items to random spots in the list
* `replace random`: replace a subset of the list at random locations
* `add start` & `add end`
* `remove start` & `remove end`
* `swap start/end`
* `sort`
* `move start/end`
* `update every 10th row`
* mode that tries to do operations similar to the [js framework benchmarks](https://krausest.github.io/js-framework-benchmark/2020/table_chrome_85.0.4183.83.html)

sorvell commented on packages/benchmarks/lit-html/repeat/index.ts:
```diff
@@ -0,0 +1,230 @@
+      }
+
+      if (s.moveCount) {
+        if (s.from < s.to) {
```
min/max from/to to save duplication?

kevinpschaaf replied:
Not sure how that would help? You need to do the slicing in a different order when moving forward vs. backward.  Unless I'm missing a clever trick you're pointing out, might need to see code.

sorvell replied:
Meh, it's readable as is.

kevinpschaaf commented on packages/benchmarks/lit-html/repeat/index.ts:
```diff
@@ -0,0 +1,230 @@
+
+const itemTemplates = {
+  li: (item: Item) => html`<li>${item.text}</li>`,
+  input: (item: Item) => html`<div>${item.text}: <input value="${item.text}"></div>`
```
Only you change `itemType` on the query string.

kevinpschaaf commented on packages/benchmarks/lit-html/repeat/index.ts:
```diff
@@ -0,0 +1,230 @@
+  repeat(items: Item[], renderItem: (item: Item) => any) {
+    return html`<ul>${repeat(items, item => item.id, renderItem)}</ul>`;
+  },
+  map(items: Item[], renderItem: (item: Item) => any) {
```
The query string allows this sort of "I'd like to specifically test xyz" use of the benchmark. The `preset` is what we run through CI.

kevinpschaaf commented on packages/benchmarks/lit-html/repeat/index.ts:
```diff
@@ -0,0 +1,230 @@
+      }
+
+      if (s.removeCount) {
+        items = items.filter((_, i) => i < s.from || i >= s.from + s.removeCount);
```
No, `repeat` doesn't care if the array itself changed or not. Once you've decided to re-render a template (based on whatever dirty check mechanism you're gating `render` on), `repeat` will always diff results from the current array against the previously stored parts.

kevinpschaaf commented on .github/workflows/benchmarks.yml:
```diff
@@ -72,6 +84,43 @@ jobs:
+      # optional benchmarks for a PR
+      # # lit-html:repeat-full
+      # - name: Benchmark lit-html/repeat-full
+      #   if: ${{ contains(github.event.pull_request.body, '[benchmarks:lit-html-repeat-full]') }}
```
My thought was that non-repeat specific changes to lit-html core might also affect `repeat` performance, so better to make it a tag you could add to any PR.

The idea was that _if_ you noticed repeat perf changing, you could add the tag to the description (not unlike how you trigger a lot of presubmit features in critique) to trigger the expanded benchmarks.

Either way, this didn't work out so hot, because edits to the metadata (and even manually re-running the workflow after that) don't cause this condition to be evaluated against the changed metadata (title or body); it's only after you push a new commit that the conditions changed, from what I empirically found, making it sort of finicky.

sorvell replied:
OK.

kevinpschaaf commented on packages/benchmarks/lit-html/repeat/index.ts:
```diff
@@ -0,0 +1,230 @@
+  .filter(s => s)
+  .map((p) => p.split('='))
+  .reduce(
+    (p: {[key: string]: any}, [k, v]) => ((p[k] = JSON.parse(v || 'true')), p),
```
Yah ok, I think I was just quoting them. This is better. 👍

kevinpschaaf commented on .github/workflows/benchmarks.yml:
```diff
@@ -3,8 +3,18 @@ name: Benchmarks
 
 jobs:
+  setup:
+    name: Setup Tachometer Reporting
```
I was experimenting with different ways to get the conditional benchmarks running, but I think it's unnecessary in our current config so I can revert it. 

I think it's only useful if you have parallel jobs running benchmarks; this ensures the comment is cleared once before any of the jobs start (and you make those jobs dependent on this having completed), afaiu. I considered separating the repeat benchmarks out into a parallel job, but it's pretty unclear to me whether parallel jobs run on the same VM or not, and I don't think we'd want parallel benchmarking going on on the same VM.

kevinpschaaf replied:
Done

kevinpschaaf commented:
@sorvell PTAL

I added a few more, here's my rationale for your suggestions:
* [x] **add random**: I added `addStripe` and `removeStripe` which add/remove items at spaced locations down the list; for benefit of reducing noise, I'm trying to avoid randomness in the benchmarks (the only one that is random is `shuffle`)
* [x] **replace random**: same rationale as above, added `replaceStripe`
* [x] **add start & add end**: added as `addEdges`
* [x] **remove start & remove end**: added as `removeEdges`
* [x] **swap start/end**: added as `swapEdges`
* [ ] **sort**: How repeat responds to sorting an array just depends on what moves the sort actually did, which should be covered by all the other operations here, so not adding this
* [ ] **move start/end**: the `moveForward` and `moveBackward` operations are very similar to this; they just start slightly inside the list; not sure there's a ton of value of having another one that starts exactly at the beginning/end
* [x] **update every 10th row**: this benchmark doesn't do any "updates" of individual item data, since there's no O(n) component of that from repeat's POV (every item gets `setValue` called on it, so this would just be measuring the cost of `NodePart:setValue`)

sorvell commented on packages/benchmarks/lit-html/repeat/index.ts:
```diff
@@ -0,0 +1,262 @@
+  .reduce(
+    // convert boolean, number, strings and default empty to true
+    (p: {[key: string]: string|boolean}, [k, v]) => 
+      (p[k] = (() => { try { return JSON.parse(v) } catch { return v || true }})(), p),
```
Let's fix this in the other benchmarks too. Pull this into a module?
